src/cockpit/channel.py

```diff
@@ -161,9 +170,43 @@ def done(self):
     def close(self, **kwargs):
         self.send_control('close', **kwargs)
 
-    def send_data(self, data: bytes) -> None:
+    def send_data(self, data: bytes) -> int:
+        """Send data and return the number of bytes left in the sender window.
+
+        The data is sent immediately, even if it's larger than the window.  In
+        general you should try not to send packets larger than
+        Channel.CHANNEL_FLOW_PING.
+
+        A positive return value is a suggested maximum size for the next write.
+
+        If the function returns zero or negative, then the sender should stop
+        sending data until do_thaw_send() is called.
+        """
         self.send_channel_data(self.channel, data)
 
+        if not self._send_pings:
+            return Channel.CHANNEL_FLOW_PING
+
+        out_sequence = self._out_sequence + len(data)
+        if self._out_sequence // Channel.CHANNEL_FLOW_PING != out_sequence // Channel.CHANNEL_FLOW_PING:
+            self.send_control(command='ping', sequence=out_sequence)
+        self._out_sequence = out_sequence
+
+        return min(self._out_window - self._out_sequence, Channel.CHANNEL_FLOW_PING)
+
+    def do_pong(self, message):
+        if not self._send_pings:  # huh?
+            logger.warning("Got wild pong on channel %s", self.channel)
+            return
+
+        self._out_window = message['sequence'] + Channel.CHANNEL_FLOW_WINDOW
+        if self._out_sequence < self._out_window:
+            self.do_thaw_send(min(self._out_window - self._out_sequence, Channel.CHANNEL_FLOW_PING))
+
+    def do_thaw_send(self, n_bytes: int) -> None:
+        """Called to indicate that the channel may now send n_bytes more bytes."""
+        pass  # temporary.  should raise NotImplementedError
```

Comment:
Forgotten to actually do that? The subclasses implement it already, apparently.

Reply:
No.  I wanted to leave it this way.  I originally had the `raise NotImplementedError` but it was crashing.  Consider that many of our channels still subclass directly `Channel` and don't implement this.

Reply:
I see. Can you please update the comment?

Reply:
Done.